### ISSUE TITLE
Calling inspect.signature with an AsyncMock instance raises an exception

### ISSUE THREAD

**Issue by yvan-the-barbarian** [python/cpython#96127]
# Bug report

Starting in Python 3.10.6, the following code raises an exception:

```python
from inspect import signature
from unittest.mock import AsyncMock

mock = AsyncMock()
signature(mock)
```

* No exception is raised when `Mock` is used.
* The exception does not occur with Python 3.10.5.

# Your environment

- CPython versions tested on: 3.10.4, 3.10.5, 3.10.6 (Only occurs in 3.10.6)
- Operating system and architecture: Windows 10, Linux (Official Docker Image)

### Linked PRs
* gh-96335
* gh-101646
* gh-101647
* gh-108652

**Comment by graingert** [comment]
Related? https://github.com/python/cpython/issues/94924

**Comment by sobolevn** [comment]
@graingert yes, your link was very useful. Please, take a look at my proposed soultion :)

**Comment by cjw296** [comment]
@sobolevn - where is your proposed solution?

**Comment by AlexWaygood** [comment]
> @sobolevn - where is your proposed solution?

- https://github.com/python/cpython/pull/96335

**Comment by hauntsaninja** [comment]
Looks like this has been fixed, thanks all!

**Comment by mdczaplicki** [comment]
I'm on 3.11.1 and this is happening.

```
    def _signature_from_function(cls, func, skip_bound_arg=True,
                                 globals=None, locals=None, eval_str=False):
        """Private helper: constructs Signature for the given python function."""
    
        is_duck_function = False
        if not isfunction(func):
            if _signature_is_functionlike(func):
                is_duck_function = True
            else:
                # If it's not a pure Python function, and not a duck type
                # of pure function:
                raise TypeError('{!r} is not a Python function'.format(func))
    
        s = getattr(func, "__text_signature__", None)
        if s:
            return _signature_fromstr(cls, func, s, skip_bound_arg)
    
        Parameter = cls._parameter_cls
    
        # Parameter information.
        func_code = func.__code__
        pos_count = func_code.co_argcount
        arg_names = func_code.co_varnames
        posonly_count = func_code.co_posonlyargcount
>       positional = arg_names[:pos_count]
E       TypeError: 'Mock' object is not subscriptable
```

**Comment by tirkarthi** [comment]
This commit is present on main branch and 3.12 latest alpha. I guess this is not backported.

**Comment by arhadthedev** [comment]
@sobolevn should the PR be backported to 3.11?

**Comment by sobolevn** [comment]
Automatic backport has failed :(
I don't have time right now to do it manually, please help me!

**Comment by arhadthedev** [comment]
Got it, will start this in a couple of hours.

**Comment by arhadthedev** [comment]
@sobolevn ping?

**Comment by arhadthedev** [comment]
@mdczaplicki @tirkarthi The backporting is done. The fix will be available in 3.11.3 and 3.10.11 this April.

**Comment by AlexWaygood** [comment]
> The fix will be available in 3.11.3 and 3.10.11 this April.

@arhadthedev where is the 3.10 backport?

**Comment by arhadthedev** [comment]
I confused it with my other issue that was backported to both versions. Thank you for pointing out, edited.
